Options.md

# `blender2ogre` Options

## Index
 - [Exporter](#exporter)
   - [Exporter Options](#exporter-options)
   - [Exporter Script](#exporter-script)
 - [Importer](#importer)
   - [Importer Options](#importer-options)
   - [Importer Script](#importer-script)

## Exporter

### Exporter Options
Option|Name|Description|Default Value
|---|---|---|---|
|**General**|
|EX_SWAP_AXIS|Swap Axis|Axis swapping mode|'xyz'|
|EX_V2_MESH_TOOL_VERSION|Mesh Export Version|Specify Ogre version format to write|'v2'|
|EX_XML_DELETE|Clean up xml files|Remove the generated xml files after binary conversion.[^1]|True|
|**Scene**|
|EX_SCENE|Export Scene|Export current scene (OgreDotScene xml file)|True|
|EX_SELECTED_ONLY|Export Selected Only|Export only selected objects. Turn on to avoid exporting non-selected stuff|True|
|EX_EXPORT_HIDDEN|Export Hidden Also|Export hidden meshes in addition to visible ones. Turn off to avoid exporting hidden stuff|True|
|EX_EXPORT_USER|Export User Properties|Export user properties such as as physical properties. Turn off to avoid exporting the user data|True|
|EX_FORCE_CAMERA|Force Camera|Export active camera|True|
|EX_FORCE_LAMPS|Force Lamps|Export all Lamps|True|
|EX_NODE_ANIMATION|Export Node Animations|Export Node Animations, these are animations of the objects properties like position, rotation and scale|True|
|**Materials**|
|EX_MATERIALS|Export Materials|Exports .material scripts|True|
|EX_SEPARATE_MATERIALS|Separate Materials|Exports a .material for each material (rather than putting all materials into a single .material file)|True|
|EX_COPY_SHADER_PROGRAMS|Copy Shader Programs|When using script inheritance copy the source shader programs to the output path|True|
|EX_USE_FFP_PARAMETERS|Fixed Function Parameters|Convert material parameters to Blinn-Phong model|False|
|**Textures**|
|EX_DDS_MIPS|DDS Mips|Number of Mip Maps (DDS)|16|
|EX_FORCE_IMAGE_FORMAT|Convert Images|Convert all textures to selected image format|'NONE'|
|**Armature**|
|EX_ARMATURE_ANIMATION|Armature Animation|Export armature animations (updates the .skeleton file)|True|
|EX_SHARED_ARMATURE|Shared Armature|Export a single .skeleton file for objects that have the same Armature parent[^2]|False|
|EX_ONLY_KEYFRAMES|Only Keyframes|Only export Keyframes.[^3]|False|
|EX_ONLY_DEFORMABLE_BONES|Only Deformable Bones|Only exports bones that are deformable.[^4]|False|
|EX_ONLY_KEYFRAMED_BONES|Only Keyframed Bones|Only exports bones that have been keyframed for a given animation. Useful to limit the set of bones on a per-animation basis|False|
|EX_OGRE_INHERIT_SCALE|OGRE Inherit Scale|Whether the OGRE bones have the 'inherit scale' flag on.[^5]|False|
|EX_TRIM_BONE_WEIGHTS|Trim Weights|Ignore bone weights below this value (Ogre supports 4 bones per vertex)|0.01|
|**Mesh Options**|
|EX_MESH|Export Meshes|Export meshes|True|
|EX_MESH_OVERWRITE|Export Meshes (overwrite)|Export meshes (overwrite existing files)|True|
|EX_ARRAY|Optimise Arrays|Optimise array modifiers as instances (constant offset only)|True|
|EX_V1_EXTREMITY_POINTS|Extremity Points|[^6]|0|
|EX_Vx_GENERATE_EDGE_LISTS|Generate Edge Lists|Generate Edge Lists (for Stencil Shadows)|False|
|EX_GENERATE_TANGENTS|Tangents|Export tangents generated by Blender[^7]|0|
|EX_Vx_OPTIMISE_ANIMATIONS|Optimise Animations|DON"T optimise out redundant tracks & keyframes|True|
|EX_V2_OPTIMISE_VERTEX_BUFFERS|Optimise Vertex Buffers For Shaders|Optimise vertex buffers for shaders.[^8]|True|
|EX_V2_OPTIMISE_VERTEX_BUFFERS_OPTIONS|Vertex Buffers Options|Used when optimizing vertex buffers for shaders.[^9]|'puqs'|
|**LOD**|
|EX_LOD_LEVELS|LOD Levels|Number of LOD levels|0|
|EX_LOD_DISTANCE|LOD Distance|Distance increment to reduce LOD|300|
|EX_LOD_PERCENT|LOD Percentage|LOD percentage reduction|40|
|EX_LOD_MESH_TOOLS|Use OgreMesh Tools|Use OgreMeshUpgrader/OgreMeshTool instead of Blender to generate the mesh LODs.[^10]|False|
|**Pose Animation**|
|EX_SHAPE_ANIMATIONS|Shape Animation|Export shape animations (updates the .mesh file)|True|
|EX_SHAPE_NORMALS|Shape Normals|Export normals in shape animations (updates the .mesh file)|True|
|**Logging**|
|EX_Vx_ENABLE_LOGGING|Write Exporter Logs|Write Log file to the output directory (blender2ogre.log)|False|
|EX_Vx_DEBUG_LOGGING|Debug Logging|Whether to show DEBUG log messages|False|

[^1]: The removal will only happen if OgreXMLConverter/OgreMeshTool finishes successfully
[^2]: This is useful for using with: `shareSkeletonInstanceWith()`
  NOTE: The name of the.skeleton file will be that of the Armature
[^3]: Exported animation won't be affected by Inverse Kinematics, Drivers and modified F-Curves
[^4]: Useful for hiding IK-Bones used in Blender.
  NOTE: Any bone with deformablechildren/descendants will be output as well
[^5]: If the animation has scale in it, the exported animation needs to be adjusted to account for the state of the inherit-scale flag in OGRE
[^6]: Submeshes can have optional "extremity points" stored with them to allow submeshes to be sorted with respect to each other in the case of transparency.
  For some meshes with transparent materials (partial transparency) this can be useful
[^7]: Options:
  '0': Do not export
  '3': Generate
  '4': Generate with parity
[^8]: See Vertex Buffers Options for more settings
[^9]: Available flags are:
  p - converts POSITION to 16-bit floats.
  q - converts normal tangent and bitangent (28-36 bytes) to QTangents (8 bytes).
  u - converts UVs to 16-bit floats.
  s - make shadow mapping passes have their own optimised buffers. Overrides existing ones if any.
  S - strips the buffers for shadow mapping (consumes less space and memory)
[^10]: OgreMeshUpgrader/OgreMeshTool does LOD by removing edges, which allows only changing the index buffer and re-use the vertex-buffer (storage efficient).
  Blenders decimate does LOD by collapsing vertices, which can result in a visually better LOD, but needs different vertex-buffers per LOD.

### Exporter Script
This is an example exporting script with all the options and their default values

```python
import bpy

bpy.ops.ogre.export(
filepath="D:\\tmp\\NormalsExport\\blender2ogre.scene", 

# General
EX_SWAP_AXIS='xz-y', 
# - 'xyz': No swapping
# - 'xz-y'OGRE Standard
# - '-xzy': Non standard
EX_V2_MESH_TOOL_VERSION='v2', 
# - 'v1': Export the mesh as a v1 object
# - 'v2': Export the mesh as a v2 object
EX_XML_DELETE=True, 

# Scene
EX_SCENE=True, 
EX_SELECTED_ONLY=True, 
EX_EXPORT_HIDDEN=True, 
EX_FORCE_CAMERA=True, 
EX_FORCE_LAMPS=True, 
EX_NODE_ANIMATION=True, 

# Materials
EX_MATERIALS=True, 
EX_SEPARATE_MATERIALS=True, 
EX_COPY_SHADER_PROGRAMS=True, 

# Textures
EX_DDS_MIPS=16, 
EX_FORCE_IMAGE_FORMAT='NONE', 

# Armature
EX_ARMATURE_ANIMATION=True, 
EX_SHARED_ARMATURE=False, 
EX_ONLY_KEYFRAMES=False, 
EX_ONLY_DEFORMABLE_BONES=False, 
EX_ONLY_KEYFRAMED_BONES=False, 
EX_OGRE_INHERIT_SCALE=False, 
EX_TRIM_BONE_WEIGHTS=0.01, 

# Mesh Options
EX_MESH=True, 
EX_MESH_OVERWRITE=True, 
EX_ARRAY=True,
EX_V1_EXTREMITY_POINTS=0, 
EX_Vx_GENERATE_EDGE_LISTS=False, 
EX_GENERATE_TANGENTS='0', 
# - '0': Do not export
# - '3': Generate
# - '4': Generate with parity
EX_Vx_OPTIMISE_ANIMATIONS=True, 
EX_V2_OPTIMISE_VERTEX_BUFFERS=True, 
EX_V2_OPTIMISE_VERTEX_BUFFERS_OPTIONS="puqs", 

# LOD
EX_LOD_LEVELS=0, 
EX_LOD_DISTANCE=300, 
EX_LOD_PERCENT=40, 
EX_LOD_MESH_TOOLS=False, 

# Pose Animation
EX_SHAPE_ANIMATIONS=True, 
EX_SHAPE_NORMALS=True, 

# Logging
EX_Vx_ENABLE_LOGGING=True,
EX_Vx_DEBUG_LOGGING=True
)
```

## Importer

### Importer Options
Option|Name|Description|Default Value
|---|---|---|---|
|**General**|
|IM_SWAP_AXIS|Swap Axis|Axis swapping mode|'xyz'|
|IM_V2_MESH_TOOL_VERSION|Mesh Import Version|Specify Ogre version format to read|'v2'|
|IM_XML_DELETE|Clean up xml files|Remove the generated xml files after binary conversion.[^11]|True|
|**Mesh**|
|IM_IMPORT_NORMALS|Import Normals|Import custom mesh normals|True|
|IM_MERGE_SUBMESHES|Merge Submeshes|Whether to merge submeshes to form a single mesh with different materials|True|
|**Armature**|
|IM_IMPORT_ANIMATIONS|Import animation|Import animations as actions|True|
|IM_ROUND_FRAMES|Adjust frame rate|Adjust scene frame rate to match imported animation|True|
|IM_USE_SELECTED_SKELETON|Use selected skeleton|Link with selected armature object rather than importing a skeleton.[^12]|True|
|**Shape Keys**|
|IM_IMPORT_SHAPEKEYS|Import shape keys|Import shape keys (morphs)|True|
|**Logging**|
|IM_Vx_ENABLE_LOGGING|Write Importer Logs|Write Log file to the output directory (blender2ogre.log)|False|

[^11]: The removal will only happen if OgreXMLConverter/OgreMeshTool finishes successfully
[^12]: Use this for importing skinned meshes that don't have their own skeleton.
  Make sure you have the correct skeleton selected or the weight maps may get mixed up.

### Importer Script
This is an example importing script with all the options and their default values

```python
import bpy

bpy.ops.ogre.import_mesh(
filepath="D:\\tmp\\NormalsExport\\Suzanne.mesh", 

# General
IM_SWAP_AXIS='xz-y',            # Axis swapping mode
IM_V2_MESH_TOOL_VERSION='v2',   # Specify Ogre version format to read
IM_XML_DELETE=True,             # Remove the generated xml files after binary conversion.

# Mesh
IM_IMPORT_NORMALS=True,         # Import custom mesh normals
IM_MERGE_SUBMESHES=True,        # Whether to merge submeshes to form a single mesh with different materials

# Armature
IM_IMPORT_ANIMATIONS=True,      # Import animations as actions
IM_ROUND_FRAMES=True,           # Adjust scene frame rate to match imported animation
IM_USE_SELECTED_SKELETON=True,  # Link with selected armature object rather than importing a skeleton

# Shape Keys
IM_IMPORT_SHAPEKEYS=True,       # Import shape keys (morphs)

# Logging
IM_Vx_ENABLE_LOGGING=True       # Write Log file to the output directory (blender2ogre.log)
)
```


CHANGELOG.txt
CHANGELOG
    0.6.2
    * pulled fixed added to the monoltic python file
    * added exporting of custom vertex groups
    * console export documented
    0.6.1
    * code refactored, tested to work with 2.71
    0.6.0
    * patched to work with 2.66.
    0.5.9
    * apply patch from Thomas for Blender 2.6x support
    0.5.8
    * Clean all names that will be used as filenames on disk. Adjust all places
      that use these names for refs instead of ob.name/ob.data.name. Replaced chars
      are \, /, :, *, ?, ", <, >, | and spaces. Tested on work with ogre 
      material, mesh and skeleton writing/refs inside the files and txml refs.
      Shows warning at final report if we had to resort to the renaming so user 
      can possibly rename the object.
    * Added silent auto update checks if blender2ogre was installed using
      the .exe installer. This will keep people up to date when new versions are out.
    * Fix tracker issue 48: Needs to check if outputting to /tmp or 
      ~/.wine/drive_c/tmp on Linux. Thanks to vax456 for providing the patch,
      added him to contributors. Preview mesh's are now placed under /tmp 
      on Linux systems if the OgreMeshy executable ends with .exe
    * Fix tracker issue 46: add operationtype to <submesh>
    * Implement a modal dialog that reports if material names have invalid
      characters and cant be saved on disk. This small popup will show until
      user presses left or right mouse (anywhere).
    * Fix tracker issue 44: XML Attributes not properly escaped in .scene file
    * Implemented reading OgreXmlConverter path from windows registry.
      The .exe installer will ship with certain tools so we can stop guessing
      and making the user install tools separately and setting up paths.
    * Fix bug that .mesh files were not generated while doing a .txml export.
      This was result of the late 2.63 mods that forgot to update object
      facecount before determining if mesh should be exported.
    * Fix bug that changed settings in the export dialog were forgotten when you
      re-exported without closing blender. Now settings should persist always
      from the last export. They are also stored to disk so the same settings
      are in use when if you restart Blender.
    * Fix bug that once you did a export, the next time the export location was
      forgotten. Now on sequential exports, the last export path is remembered in
      the export dialog.
    * Remove all local:// from asset refs and make them relative in .txml export.
      Having relative refs is the best for local preview and importing the txml
      to existing scenes.
    * Make .material generate what version of this plugins was used to generate
      the material file. Will be helpful in production to catch things.
      Added pretty printing line endings so the raw .material data is easier to read.
    * Improve console logging for the export stages. Run Blender from
      cmd prompt to see this information.
    * Clean/fix documentation in code for future development
    * Add todo to code for future development
    * Restructure/move code for easier readability
    * Remove extra white spaces and convert tabs to space
    0.5.7
    * Update to Blender 2.6.3.
    * Fixed xz-y Skeleton rotation (again)
    * Added additional Keyframe at the end of each animation to prevent
      ogre from interpolating back to the start
    * Added option to ignore non-deformable bones
    * Added option to export nla-strips independently from each other

TODO
    * Remove this section and integrate below with code :)
    * Fix terrain collision offset bug
    * Add realtime transform (rotation is missing)
    * Fix camera rotated -90 ogre-dot-scene
    * Set description field for all pyRNA


io_ogre/config.py
import bpy, os, sys, logging, pickle, mathutils
from pprint import pprint
from bpy.props import *

logger = logging.getLogger('config')

AXIS_MODES =  [
    ('xyz', 'xyz', 'No Axis swapping'),
    ('xz-y', 'xz-y', 'Ogre standard'),
    ('-xzy', '-xzy', 'Non standard'),
]

MESH_TOOL_VERSIONS = [
    ('v1', 'v1', 'Export the mesh as a v1 object'),
    ('v2', 'v2', 'Export the mesh as a v2 object')
]

TANGENT_MODES =  [
    ('0', 'none', 'Do not export tangents'),
    ('3', 'generate', 'Generate tangents'),
    ('4', 'generate with parity', 'Generate with parity')
]

CONFIG_PATH = bpy.utils.user_resource('CONFIG', path='scripts', create=True)
CONFIG_FILENAME = 'io_ogre.pickle'
CONFIG_FILEPATH = os.path.join(CONFIG_PATH, CONFIG_FILENAME)

_CONFIG_DEFAULTS_ALL = {
    # General
    'SWAP_AXIS' : 'xyz', # ogre standard is 'xz-y', but swapping is currently broken
    'MESH_TOOL_VERSION' : 'v2',
    'XML_DELETE' : True,

    # Scene
    'SCENE' : True,
    'SELECTED_ONLY' : True,
    'EXPORT_HIDDEN' : True,
    #'EXPORT_USER' : True,
    'FORCE_CAMERA' : True,
    'FORCE_LAMPS' : True,
    'NODE_ANIMATION' : True,
    #'NODE_KEYFRAMES' : False,

    # Materials
    'MATERIALS' : True,
    'COPY_SHADER_PROGRAMS' : True,
    'SEPARATE_MATERIALS' : True,
    'USE_FFP_PARAMETERS': False,

    # Textures
    'MAX_TEXTURE_SIZE' : 4096,
    'FORCE_IMAGE_FORMAT' : 'NONE',
    'TOUCH_TEXTURES' : True,
    'DDS_MIPS' : 16,

    # Armature
    'ONLY_DEFORMABLE_BONES' : False,
    'ONLY_KEYFRAMED_BONES' : False,
    'OGRE_INHERIT_SCALE' : False,
    'ARMATURE_ANIMATION' : True,
    'TRIM_BONE_WEIGHTS' : 0.01,
    'ONLY_KEYFRAMES' : False,
    'SHARED_ARMATURE' : False,

    # Mesh
    'MESH' : True,
    'MESH_OVERWRITE' : True,
    'ARRAY' : True,
    'EXTREMITY_POINTS' : 0,
    'GENERATE_EDGE_LISTS' : False,
    'GENERATE_TANGENTS' : "0",
    'OPTIMISE_ANIMATIONS' : True,
    'INTERFACE_TOGGLE': False,
    'OPTIMISE_VERTEX_BUFFERS' : True,
    'OPTIMISE_VERTEX_BUFFERS_OPTIONS' : 'puqs',

    # LOD
    'LOD_LEVELS' : 0,
    'LOD_DISTANCE' : 300,
    'LOD_PERCENT' : 40,
    'LOD_MESH_TOOLS' : False,

    # Pose Animation
    'SHAPE_ANIMATIONS' : True,
    'SHAPE_NORMALS' : True,

    # Logging
    'ENABLE_LOGGING' : False,
    'DEBUG_LOGGING' : False,
    #'SHOW_LOG_NAME' : False,

    # Import
    'IMPORT_NORMALS' : True,
    'MERGE_SUBMESHES' : True,
    'IMPORT_ANIMATIONS' : True,
    'ROUND_FRAMES' : True,
    'USE_SELECTED_SKELETON' : True,
    'IMPORT_SHAPEKEYS' : True
}

_CONFIG_TAGS_ = 'OGRETOOLS_XML_CONVERTER OGRETOOLS_MESH_UPGRADER MESH_PREVIEWER IMAGE_MAGICK_CONVERT USER_MATERIALS SHADER_PROGRAMS'.split()

''' todo: Change pretty much all of these windows ones. Make a smarter way of detecting
    Ogre tools from various default folders. Also consider making a installer that
    ships Ogre cmd line tools to ease the setup steps for end users. '''

_CONFIG_DEFAULTS_WINDOWS = {
    'OGRETOOLS_XML_CONVERTER' : 'C:\\OgreCommandLineTools\\OgreXMLConverter.exe',
    'OGRETOOLS_MESH_UPGRADER' : 'C:\\OgreCommandLineTools\\OgreMeshUpgrader.exe',
    'MESH_PREVIEWER' : 'ogre-meshviewer.bat',
    'IMAGE_MAGICK_CONVERT' : 'C:\\Program Files\\ImageMagick\\convert.exe',
    'USER_MATERIALS' : '',
    'SHADER_PROGRAMS' : 'C:\\'
}

_CONFIG_DEFAULTS_UNIX = {
    # do not use absolute paths like /usr/bin/exe_name. some distris install to /usr/local/bin ...
    # just trust the env PATH variable
    'IMAGE_MAGICK_CONVERT' : 'convert',
    'OGRETOOLS_XML_CONVERTER' : 'OgreXMLConverter',
    'OGRETOOLS_MESH_UPGRADER' : 'OgreMeshUpgrader',
    'MESH_PREVIEWER' : 'ogre-meshviewer',
    'USER_MATERIALS' : '',
    'SHADER_PROGRAMS' : '~/',
    #'USER_MATERIALS' : '~/ogre_src_v1-7-3/Samples/Media/materials',
    #'SHADER_PROGRAMS' : '~/ogre_src_v1-7-3/Samples/Media/materials/programs',
}

# Unix: Replace ~ with absolute home dir path
if sys.platform.startswith('linux') or sys.platform.startswith('darwin') or sys.platform.startswith('freebsd'):
    for tag in _CONFIG_DEFAULTS_UNIX:
        path = _CONFIG_DEFAULTS_UNIX[ tag ]
        if path.startswith('~'):
            _CONFIG_DEFAULTS_UNIX[ tag ] = os.path.expanduser( path )
        elif tag.startswith('OGRETOOLS') and not os.path.isfile( path ):
            _CONFIG_DEFAULTS_UNIX[ tag ] = os.path.join( '/usr/bin', os.path.split( path )[-1] )
    del tag
    del path


## PUBLIC API continues

def load_config():
    config_dict = {}

    # Check if the config file exists and load it
    if os.path.isfile( CONFIG_FILEPATH ):
        try:
            with open( CONFIG_FILEPATH, 'rb' ) as f:
                config_dict = pickle.load( f )
        except:
            logger.error('Can not read config from %s' % CONFIG_FILEPATH)

    # Load default values from _CONFIG_DEFAULTS_ALL if they don't exist after loading config from file
    for tag in _CONFIG_DEFAULTS_ALL:
        if tag not in config_dict:
            config_dict[ tag ] = _CONFIG_DEFAULTS_ALL[ tag ]

    # Load default values from _CONFIG_DEFAULTS_WINDOWS or _CONFIG_DEFAULTS_UNIX if they don't exist after loading config from file
    for tag in _CONFIG_TAGS_:
        if tag not in config_dict:
            if sys.platform.startswith('win'):
                config_dict[ tag ] = _CONFIG_DEFAULTS_WINDOWS[ tag ]
            elif sys.platform.startswith('linux') or sys.platform.startswith('darwin') or sys.platform.startswith('freebsd'):
                config_dict[ tag ] = _CONFIG_DEFAULTS_UNIX[ tag ]
            else:
                logger.error('Unknown platform: %s' % sys.platform)
                assert 0

    # Setup temp hidden RNA to expose the file paths
    for tag in _CONFIG_TAGS_:
        default = config_dict[ tag ]
        #func = eval( 'lambda self,con: config_dict.update( {"%s" : self.%s} )' %(tag,tag) )
        func = lambda self,con: config_dict.update( {tag : getattr(self,tag,default)} )
        if type(default) is bool:
            prop = BoolProperty( name=tag,
                                 description='updates bool setting',
                                 default=default,
                                 options={'SKIP_SAVE'},
                                 update=func)
        else:
            prop = StringProperty( name=tag,
                    description='updates path setting',
                    maxlen=128,
                    default=default,
                    options={'SKIP_SAVE'},
                    update=func)
        setattr( bpy.types.WindowManager, tag, prop )

    return config_dict

# Global CONFIG dictionary
CONFIG = load_config()

def get(name, default=None):
    global CONFIG
    if name in CONFIG:
        return CONFIG[name]
    return default

def update(**kwargs):
    for k,v in kwargs.items():
        if k not in _CONFIG_DEFAULTS_ALL:
            logger.warn("Trying to set CONFIG['%s'] = %s, but it is not a known config setting" % (k,v))
        CONFIG[k] = v
    save_config()

def save_config():
    global CONFIG
    #for key in CONFIG: print( '%s = %s' %(key, CONFIG[key]) )
    if os.path.isdir( CONFIG_PATH ):
        try:
            with open( CONFIG_FILEPATH, 'wb' ) as f:
                pickle.dump( CONFIG, f, -1 )
        except:
            logger.error('Can not write to %s' % CONFIG_FILEPATH)
    else:
        logger.error('Config directory %s does not exist' % CONFIG_PATH)

def update_from_addon_preference(context):
    addon_preferences = context.preferences.addons["io_ogre"].preferences

    for key in _CONFIG_TAGS_:
        addon_pref_value = getattr(addon_preferences,key,None)
        if addon_pref_value is not None:
            CONFIG[key] = addon_pref_value
